source/extensions/transport_sockets/s2a/config.cc

```
@@ -0,0 +1,119 @@
+  auto config =
+      MessageUtil::downcastAndValidate<const envoy::extensions::transport_sockets::s2a::v3alpha::S2AConfiguration&>(
+          message, factory_ctxt.messageValidationVisitor());
+
+  const std::string& s2a_address = config.s2a_address();
+  HandshakerFactory factory =
+      [s2a_address, is_upstream,
+       s2a_shared_state](Event::Dispatcher& dispatcher,
+                          const Network::Address::InstanceConstSharedPtr& local_address,
+                          const Network::Address::InstanceConstSharedPtr&) -> TsiHandshakerPtr {
+    ASSERT(local_address != nullptr);
+
+    grpc_s2a_credentials_options* options = grpc_s2a_credentials_options_create();
+    grpc_s2a_credentials_options_set_s2a_address(options, s2a_address.c_str());
+    const char* target_name = is_upstream ? "" : nullptr;
+    s2a::tsi::S2ATsiHandshakerOptions s2a_options{/* interested_parties= */nullptr, is_upstream, options, target_name};
+    // Specifying target name as empty since TSI won't take care of validating peer identity
+    // in this use case. The validation will be performed by TsiSocket with the validator.
+    // Set the max frame size to 16KB.
+    absl::StatusOr<tsi_handshaker*> tsi_handshaker_ptr = CreateS2ATsiHandshaker(s2a_options);
+
+    if (!tsi_handshaker_ptr.ok()) {
+      const std::string handshaker_name = is_upstream ? "client" : "server";
+      ENVOY_LOG_MISC(warn, "Cannot create S2A {} handshaker, status: {}", handshaker_name, tsi_handshaker_ptr.status());
+      return nullptr;
+    }
+
+    CHandshakerPtr handshaker_ptr{*tsi_handshaker_ptr};
+
+    return std::make_unique<TsiHandshaker>(std::move(*tsi_handshaker_ptr), dispatcher);
+  };
+
+  return std::make_unique<TsiSocketFactory>(factory, /* validator= */ nullptr);
+}
+
+} // namespace
+
+ProtobufTypes::MessagePtr S2ATransportSocketConfigFactory::createEmptyConfigProto() {
+  return std::make_unique<envoy::extensions::transport_sockets::s2a::v3alpha::S2AConfiguration>();
```

I'm a little confused why we're returning in this case rather than e.g. logging an error and returning a nullptr. I'm not sure if ALTS has the HS address baked into the code, but for S2A we should fail immediately if the S2A address is not set.

---

Oh I think I understand - this is for testing only? If that's the case (and the transport socket API that we're overriding agrees with that), please add a comment in the .h that explains that.